the-collab-lab/tcl-12-smart-shopping-list
# #6: Alert user when adding items with a name that already exists

## Description
This update alerts the user when they try to add an item with a duplicate name to their shopping list. This also accounts for capitalization and non-alphanumerical characters.
- Restructured the Firestore data to store items with their normalized names as keys and the item details as the value
- Refactored the FirestoreList component to account for different data structure
- Added a duplicate item name check when adding an item to the list
- Added error handling for Firestore updates and duplicate checks

## Related Issue
Closes #6 

## Acceptance Criteria

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓ | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### After
Firestore Structure:
![Screen Shot 2020-10-20 at 7 20 55 PM](https://user-images.githubusercontent.com/39573970/96655218-d252eb80-12f1-11eb-966e-15e10db3c14f.png)
---
Duplicate item name message:
![Screen Shot 2020-10-20 at 7 21 28 PM](https://user-images.githubusercontent.com/39573970/96655228-d7b03600-12f1-11eb-8d95-81c65a8026dc.png)
---
Duplicate item name accounts for non-alphanumeric characters:
![Screen Shot 2020-10-20 at 7 21 49 PM](https://user-images.githubusercontent.com/39573970/96655237-daab2680-12f1-11eb-8c78-f6a821fe114b.png)
---
Error message for database errors:
![Screen Shot 2020-10-20 at 7 26 48 PM](https://user-images.githubusercontent.com/39573970/96655251-e0087100-12f1-11eb-8fad-e9f9213b5ef4.png)

## Testing Steps / QA Criteria
- Pull latest from branch `al-tc-avoid-dupe-items` and start server
- Create a list and add an item to the list
- Add the item again to see the error message
- Add special characters to the item name to see error messages again
- Change `db.collection('lists')` (see comment) to see database error messages
